### PR TITLE
🔀 Added Cache Support

### PR DESCRIPTION
**Pull request by furkandeveloper**
## Summary

Hi, I developed library for cache operations. This library of about see for more information. [EasyCache](https://github.com/furkandeveloper/EasyCache)

### Purpose of EasyCache
EasyCache includes multiple cache providers. So can be easily change cache provider. Also EasyCache can be customized. 
See wiki for technical information. [Wiki](https://github.com/furkandeveloper/EasyCache/wiki)

### Configuration
EasyCache configuration is similar to database configuration in CleanArchitecture.

```json
"CacheOptions": {
    "IsActive": true,
    "ProviderName": "Memory",
    "Redis": {
      "Configuration": "localhost",
      "InstanceName": "CleanArchitecture"
    },
    "Memory": {
    },
    "MemCache": {
      "ServerName": "localhost",
      "Port": 11211
    }
  }
```

```csharp
if (configuration.GetValue<bool>(Constants.CACHE_OPTIONS_IS_ACTIVE))
            {
                switch (configuration.GetValue<string>(Constants.CACHE_OPTIONS_PROVIDER_NAME))
                {
                    case Constants.CACHE_REDIS:
                        RedisCacheOption redisCacheOption = new RedisCacheOption();
                        configuration.GetSection(Constants.CACHE_OPTIONS_REDIS).Bind(redisCacheOption, c => c.BindNonPublicProperties = true);
                        services.AddEasyRedisCache(options =>
                        {
                            options.Configuration = redisCacheOption.Configuration;
                            options.InstanceName = redisCacheOption.InstanceName;
                        });
                        break;
                    case Constants.CACHE_MEMORY:
                        services.AddEasyMemoryCache();
                        break;
                    case Constants.CACHE_MEMCACHE:
                        MemCacheOption memCacheOption = new MemCacheOption();
                        configuration.GetSection(Constants.CACHE_OPTIONS_MEMCACHE).Bind(memCacheOption, c => c.BindNonPublicProperties = true);
                        services.AddEasyMemCache(options => options.AddServer(memCacheOption.ServerName, memCacheOption.Port));
                        break;
                    default:
                        break;
                }
            }
```

### Usage
Get `IEasCacheService` from dependency injection.
```csharp
public class GetTodoItemsWithPaginationQueryHandler : IRequestHandler<GetTodoItemsWithPaginationQuery, PaginatedList<TodoItemDto>>
    {
        private readonly IApplicationDbContext _context;
        private readonly IMapper _mapper;
        private readonly IEasyCacheService _easyCacheService;

        public GetTodoItemsWithPaginationQueryHandler(IApplicationDbContext context, IMapper mapper, IEasyCacheService easyCacheService)
        {
            _context = context;
            _mapper = mapper;
            _easyCacheService = easyCacheService;
        }
}
```

```csharp
public async Task<PaginatedList<TodoItemDto>> Handle(GetTodoItemsWithPaginationQuery request, CancellationToken cancellationToken)
        {
            return await _easyCacheService.GetAndSetAsync(request.ListId.ToString(), async () => await GetTodoItemAsync(request), TimeSpan.FromMinutes(1));
        }

        private async Task<PaginatedList<TodoItemDto>> GetTodoItemAsync(GetTodoItemsWithPaginationQuery request)
        {
            return await _context.TodoItems
                .Where(x => x.ListId == request.ListId)
                .OrderBy(x => x.Title)
                .ProjectTo<TodoItemDto>(_mapper.ConfigurationProvider)
                .PaginatedListAsync(request.PageNumber, request.PageSize);
        }
```

### CODE REVIEW

**Comment by furkandeveloper** [comment]
Changed dependency injection scenario.

Sample code block;
```csharp
if (configuration.GetValue<bool>(Constants.CACHE_OPTIONS_IS_ACTIVE))
            {
                switch (configuration.GetValue<string>(Constants.CACHE_OPTIONS_PROVIDER_NAME))
                {
                    case Constants.CACHE_REDIS:
                        RedisCacheOption redisCacheOption = new RedisCacheOption();
                        configuration.GetSection(Constants.CACHE_OPTIONS_REDIS).Bind(redisCacheOption, c => c.BindNonPublicProperties = true);
                        services.AddEasyRedisCache(options =>
                        {
                            options.Configuration = redisCacheOption.Configuration;
                            options.InstanceName = redisCacheOption.InstanceName;
                        });
                        services.AddTransient<IApplicationCacheService, ApplicationEasyRedisCacheManager>();
                        break;
                    case Constants.CACHE_MEMORY:
                        services.AddEasyMemoryCache();
                        services.AddTransient<IApplicationCacheService, ApplicationEasyMemoryCacheManager>();
                        break;
                    case Constants.CACHE_MEMCACHE:
                        MemCacheOption memCacheOption = new MemCacheOption();
                        configuration.GetSection(Constants.CACHE_OPTIONS_MEMCACHE).Bind(memCacheOption, c => c.BindNonPublicProperties = true);
                        services.AddEasyMemCache(options => options.AddServer(memCacheOption.ServerName, memCacheOption.Port));
                        services.AddTransient<IApplicationCacheService, ApplicationEasyMemCacheManager>();
                        break;
                    default:
                        break;
                }
            }
```

**Review by gokhanyildiz95 — APPROVED** [review]

**Review by seydaazenler — APPROVED** [review]

**Review by enisn — APPROVED** [review]

File: src/Application/Common/Extensions/ApplicationCacheServiceExtension.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            if (data == null)
+            {
+                data = getResult();
+                applicationCacheService.Set(key, data, expireTime);
+            }
+
+            return data;
+        }
+
+        public static T GetAndSetAsync<T>(this IApplicationCacheService applicationCacheService, string key, Func<T> getResult, TimeSpan expireTime)
```
Comment (enisn) [inline src/Application/Common/Extensions/ApplicationCacheServiceExtension.cs:25]: I can't see anything Async here except method name.

```suggestion
        public static Task<T> GetAndSetAsync<T>(this IApplicationCacheService applicationCacheService, string key, Func<Task<T>> getResult, TimeSpan expireTime)
```

File: src/Infrastructure/Caching/ApplicationEasyRedisCacheManager.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+namespace CleanArchitecture.Infrastructure.Caching
+{
+    public class ApplicationEasyRedisCacheManager : EasyCacheRedisManager, IApplicationCacheService
+    {
+        public ApplicationEasyRedisCacheManager(IDistributedCache distributedCache) : base(distributedCache)
+        {
+        }
+
+        public override T Get<T>(string key)
```
Comment (enisn) [inline src/Infrastructure/Caching/ApplicationEasyRedisCacheManager.cs:18]: Why do you override if you don't change anything?

File: src/Infrastructure/Infrastructure.csproj [hunk 3]
```diff
@@ -8,6 +8,9 @@
 
   <ItemGroup>
     <PackageReference Include="CsvHelper" Version="15.0.10" />
+    <PackageReference Include="EasyCache.MemCache" Version="1.0.0" />
+    <PackageReference Include="EasyCache.Memory" Version="1.0.0" />
+    <PackageReference Include="EasyCache.Redis" Version="1.0.0" />
```
Comment (enisn) [inline src/Infrastructure/Infrastructure.csproj:13]: Is adding all dependencies good soluton? Even rest of them aren't used.

File: src/Application/Common/Extensions/ApplicationCacheServiceExtension.cs (hunk 1 above)
Comment (furkandeveloper) [reply src/Application/Common/Extensions/ApplicationCacheServiceExtension.cs]: Can you check commits @enisn 5d6f5940de9969e00dedcce117b20cbf59cfe4f8 f3fddf580b797c1adbca7ce1acf813339a7f7338 ?

File: src/Infrastructure/Caching/ApplicationEasyRedisCacheManager.cs (hunk 2 above)
Comment (furkandeveloper) [reply src/Infrastructure/Caching/ApplicationEasyRedisCacheManager.cs]: You are right @enisn. Can you check commit 445f43218a3465130622cd39569ad9640aa46077 ?

File: src/Infrastructure/Infrastructure.csproj (hunk 3 above)
Comment (furkandeveloper) [reply src/Infrastructure/Infrastructure.csproj:13]: It needs to be an example for all providers. I took as an example the method of implementing InMemory and SQL Server in CleanArchitecture. @enisn

**Review by furkandeveloper — COMMENTED** [review]
Approved

**Review by anilguleroglu — APPROVED** [review]

**Review by enisn — COMMENTED** [review]
Looks good but I couldn't see any documentation about how to use caching with clean architecture. Maybe a documentation would be better.

File: src/Application/Common/Extensions/ApplicationCacheServiceExtension.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            return data;
+        }
+
+        public static Task<T> GetAndSetAsync<T>(this IApplicationCacheService applicationCacheService, string key, Func<Task<T>> getResult, TimeSpan expireTime)
+        {
+            var data = applicationCacheService.Get<T>(key);
+
+            if (data == null)
+            {
+                data = getResult().GetAwaiter().GetResult();
```
Comment (deceptiveSimplicity) [inline src/Application/Common/Extensions/ApplicationCacheServiceExtension.cs:31]: Why not using async/await here?

**Comment by deceptiveSimplicity** [comment]
I don't like the idea of integrating such a young lib into this general template.

File: src/Application/Common/Extensions/ApplicationCacheServiceExtension.cs (hunk 4 above)
Comment (furkandeveloper) [reply src/Application/Common/Extensions/ApplicationCacheServiceExtension.cs:31]: Can be use but GetAwaiter extension already get awaitable object. @deceptiveSimplicity

**Comment by MCCshreyas** [comment]
Cashing is cross cutting concern and it should be outside of Application layer or around it.

Taking dependency of cache service interface in RequestHandler isn't good from my point of view!

File: src/Application/Common/Extensions/ApplicationCacheServiceExtension.cs (hunk 4 above)
Comment (deceptiveSimplicity) [reply src/Application/Common/Extensions/ApplicationCacheServiceExtension.cs:31]: `.GetAwaiter().GetResult()` will wait synchronous, blocking the current thread. That's why I am asking.

Comment (furkandeveloper) [reply src/Application/Common/Extensions/ApplicationCacheServiceExtension.cs]: You are right. Can you check commit 048a72b4e5aa922d6ab34f96297c07b3a9b4ed96 @deceptiveSimplicity ?

**Comment by furkandeveloper** [comment]
> Caching is a cross-cutting concern and it should be outside of the Application layer or around it.
> Taking dependency of cache service interface in `RequestHandler` isn't good from my point of view!
> 
> And adding a library dependency in the project template isn't good, because cashing is optional and it depends upon the type of project and type of domain, not everyone needs it.
> 
> But still, if someone wanted they can go with their own choice of the library, or if one doesn't want any external library they can put their implementation in the project without any library!
> You can see my implementation (#110) and you can refer to @jasontaylordev thoughts as well about the implementation!
> 
> And also I don't see anything about "Cache Invalidation" here!?

Thanks for feedbacks. You are right. Can be cache operation active/passive for application. Therefore, there is an `IsActive` parameter in the application settings. 

### Sample
```json
"CacheOptions": {
    "IsActive": true, <-- Cache Invalidation
    "ProviderName": "Memory",
    "Redis": {
      "Configuration": "localhost",
      "InstanceName": "CleanArchitecture"
    },
    "Memory": {
    },
    "MemCache": {
      "ServerName": "localhost",
      "Port": 11211
    }
  }
```

**Comment by MCCshreyas** [comment]
@furkandeveloper that's not cache invalidation. Please see discussion over my implementation you will understand!

**Comment by jasontaylordev** [comment]
Thanks for this PR, however I am not looking to add caching support at this stage.
The template is primarily focused on Clean Architecture with .NET Core, and I don't have capacity to add and support a lot of features.
Apologies that it took me so long to respond.
Happy coding! 😀

**Comment by neozhu** [comment]
@furkandeveloper 
You can consider implementing caching mechanisms with this
···
public class CachingBehaviour<TRequest, TResponse> : IPipelineBehavior<TRequest, TResponse>
    where TRequest : IRequest<TResponse> ,ICacheable
{
    private readonly IAppCache _cache;
    private readonly ILogger<CachingBehaviour<TRequest, TResponse>> _logger;

    public CachingBehaviour(
        IAppCache cache,
        ILogger<CachingBehaviour<TRequest, TResponse>> logger
        )
    {
        _cache = cache;
        _logger = logger;
    }
    public async Task<TResponse> Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate<TResponse> next)
    {
        _logger.LogTrace("{Name} is caching with {@Request}.", nameof(request),request);
        var response = await _cache.GetOrAddAsync(
            request.CacheKey,
            async () =>
            await next(),
            request.Options);
     
        return response;
    }
}
public class CacheInvalidationBehaviour<TRequest, TResponse> : IPipelineBehavior<TRequest, TResponse>
      where TRequest : IRequest<TResponse>,ICacheInvalidator 
{
    private readonly IAppCache _cache;
    private readonly ILogger<CacheInvalidationBehaviour<TRequest, TResponse>> _logger;

    public CacheInvalidationBehaviour(
        IAppCache cache,
        ILogger<CacheInvalidationBehaviour<TRequest, TResponse>> logger
        )
    {
        _cache = cache;
        _logger = logger;
    }
    public async Task<TResponse> Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate<TResponse> next)
    {
        _logger.LogTrace("{Name} cache expire with {@Request}.", nameof(request), request);
        var response = await next();
        if (!string.IsNullOrEmpty(request.CacheKey))
        {
            _cache.Remove(request.CacheKey);
        }
        request.SharedExpiryTokenSource?.Cancel();
        return response;
    }
}

**Pull request closed without merging**